include/transmem/transmem.h

```diff
@@ -153,73 +193,78 @@ friend class GraphMLWriter;
-     * @brief Returns the transformation between @a srcFrame at time @a tstamp1 to @a destFrame at time @a tstamp2.@n The query requires a third frame @a fixFrame which is
-     * stationary relative to @a srcFrame and @a destFrame.
-     * @see TransMem::getLink(const FrameID &srcFrame, const FrameID &destFrame, const Timestamp &tstamp) const
-     * @param srcFrame identifier of the source frame
-     * @param fixFrame identifier of the fix frame
-     * @param destFrame identifier of the destination frame
-     * @param tstamp1 timestamp when the transformation from the source to the fix frame is valid
-     * @param tstamp2 timestamp when the transformation from the fix frame to the destination frame is valid
-     * @return matrix encoding the requested transformation
-     * @throws InvalidArgumentException
-     * @throws NoSuchLinkFoundException
-     */
-    QMatrix4x4 getLink(const FrameID &srcFrame, const FrameID &fixFrame, const FrameID &destFrame, const Timestamp &tstamp1, const Timestamp &tstamp2) const;
-
-    /**
-     * @fn QQuaternion getBestLink(const FrameID &srcFrame, const FrameID &destFrame, Timestamp &tstamp) const
-     * @brief Returns the best transformation betwen @a srcFrame and @a destFrame.@n@n
-     * Best transformation:@n
+     * @fn StampedAndRatedTransformation getBestLink(const FrameID &srcFrame, const FrameID &destFrame, Timestamp &tstamp) const
+     * @brief Returns the best transformation betwen @a srcFrame and @a destFrame encoded in a StampedAndRatedTransformation object.@n@n
+     * Best transformatio:@n
      * Let L1,..,Ln be the links required for the transformation from @a srcFrame to @a destFrame.@n
      * Let Eij be the j-th transformaton entry inserted at time tij in the link Li.@n
      * The best transformation is the transformation obtained at time t, were t minimizes the sum over all |t-tij| from 1 to n.@n@n
      * The resolution of t is 10 ms.
-     * @param srcFrame identifier of the source frame
-     * @param destFrame identifier of the destination frame
-     * @param tstamp returns at what time the best transformation was available
-     * @return matrix encoding the requested transformation
+     * Transmem caches the entries. That is, if no link of a certain transformation was updated, the cached transformation is returned.
+     * @param srcFrame Identifier of the source frame.
+     * @param destFrame Identifier of the destination frame.
+     * @param tstamp Returns at what time the best transformation was available.
+     * @return A StampedAndRatedTransformation object encoding the requested transformation and additional quality information.
      * @throws InvalidArgumentException
      * @throws NoSuchLinkFoundException
      */
-    QMatrix4x4 getBestLink(const FrameID &srcFrame, const FrameID &destFrame, Timestamp &tstamp) const;
+    StampedAndRatedTransformation getBestLink(const FrameID &srcFrame, const FrameID &destFrame);
```

Comment:
I understand your point of using mutable but i don't understand why we have to protect the mutable variable with an extra mutex. Every public function has to acquire the lock first, no matter if it is just a query or an insertion.